ngraph/test/util/unary_test.hpp

```diff
@@ -0,0 +1,293 @@
+                      const std::initializer_list<Data<eleType>>& expeted,
+                      Tolerance tol = {})
+            {
+                if (m_function->is_dynamic())
+                {
+                    auto test_case =
+                        ngraph::test::TestCase<TestEngine, ngraph::test::TestCaseType::DYNAMIC>(
+                            m_function);
+                    do_test<eleType>(test_case, inputs, expeted, tol);
+                }
+                else
+                {
+                    auto test_case =
+                        ngraph::test::TestCase<TestEngine, ngraph::test::TestCaseType::STATIC>(
+                            m_function);
+                    do_test<eleType>(test_case, inputs, expeted, tol);
+                }
+            }
+
+            // SISO: single inputs, single output
+            template <ngraph::element::Type_t eleType = et>
+            void test(const Data<eleType>& input, const Data<eleType>& expeted, Tolerance tol = {})
+            {
+                test<eleType>(
+                    std::initializer_list<Data<eleType>>{input}, std::initializer_list<Data<eleType>>{expeted}, tol);
+            }
+
+            // MISO: multiple inputs, single output
+            template <ngraph::element::Type_t eleType = et>
+            void test(const std::initializer_list<Data<eleType>>& inputs,
+                      const Data<eleType>& expeted,
+                      Tolerance tol = {})
+            {
+                test<eleType>(inputs, std::initializer_list<Data<eleType>>{expeted}, tol);
+            }
+
+            // this overload supports passing a predictor with overloaded i/o types
+            template <ngraph::element::Type_t eleType = et,
+                      typename T = ngraph::fundamental_type_for<eleType>>
+            void test(const std::initializer_list<Data<eleType>>& inputs,
+                      T (*elewise_predictor)(T),
+                      Tolerance tol = {})
+            {
+                auto first = inputs.begin();
+                Data<eleType> expeted = *first;
+
+                for (size_t i = 0; i < first->value.size(); i++)
+                    expeted.value[i] = elewise_predictor(first->value[i]);
+
+                test<eleType>(inputs, expeted, tol);
+            }
+
+            // this overload supports passing a lambda as predictor
+            template <ngraph::element::Type_t eleType = et, typename Predictor>
+            void test(const std::initializer_list<Data<eleType>>& inputs,
+                      Predictor elewise_predictor,
+                      Tolerance tol = {})
+            {
+                auto first = inputs.begin();
+                Data<eleType> expeted = *first;
+
+                for (size_t i = 0; i < first->value.size(); i++)
+                    expeted.value[i] = elewise_predictor(first->value[i]);
+
+                test<eleType>(inputs, expeted, tol);
+            }
```

Comment:
use `std::function` to rule them all
```suggestion
            // this overload supports passing a predictor with overloaded i/o types
            template <ngraph::element::Type_t eleType = et,
                      typename T = ngraph::fundamental_type_for<eleType>>
            void test(const std::initializer_list<Data<eleType>>& inputs,
                      std::function<T(T)> elewise_predictor,
                      Tolerance tol = {})
            {
                auto first = inputs.begin();
                Data<eleType> expeted = *first;

                for (size_t i = 0; i < first->value.size(); i++)
                    expeted.value[i] = elewise_predictor(first->value[i]);

                test<eleType>(inputs, expeted, tol);
            }
```

Reply:
Sorry, I tried, std::function<T(T)> cannot handle overloaded function. When caller pass std::sin, compiler can choose the correct overload if you specify elewise_predictor to be function pointer with explicit prototype, but std::function's constructor is too flexible and compiler don't know which overload of std::sin to use.

On the other hand, we also want to be able to allow caller to pass a general callable (like lambda), in this case, function pointer version cannot work, so we also provided a generic callable version of elewise_predictor. And compiler will choose function pointer version if possible, then it may fallback to generic callable version.

Reply:
What if you wrap `std::sin` in lambda `[](float f){return std::sin(f);}` ?

Reply:
That certainly works for sure, but lambda is limited in C++11, you have to specify exact argument type, which is another source of bug.

Besides overloaded function like std::sin, when implementing predictor, we prefer generic function template. for example, following function template f_mish() can be used to predict result of Mish op for any input data types:

```
template <typename T>
T f_mish(T x)
{
    return x * std::tanh(std::log(1.0 + std::exp(x)));
}
```

so later on we can use the same predictor for f32/f16 tests in a very clean and less error-prone generic way:

```
make_mish_test<element::f32>(shape2).test({rand_generator(), shape2}, f_mish, 1e-5f);
...
make_mish_test<element::f16>(shape1).test({rand_generator(), shape1}, f_mish);
```

But is we use std::function<T(T)>, compiler would fail to resolve the template and we have to change the code into:

```
make_mish_test<element::f32>(shape2).test({rand_generator(), shape2}, f_mish<float>, 1e-5f);
...
make_mish_test<element::f16>(shape1).test({rand_generator(), shape1}, f_mish<float16>);
```

or more verbose:
```
make_mish_test<element::f32>(shape2).test({rand_generator(), shape2}, [](float x){ return f_mish(x);}, 1e-5f);
...
make_mish_test<element::f16>(shape1).test({rand_generator(), shape1}, [](float16 x){ return f_mish(x);});
```
This requires caller to know the exact fundamental type for f32/f16 in advance, more terrible, compiler doesn't check whether you gives correct type, you can give f_mish<int> to std::function<T(T)> when T is float, compiler complains nothing !

Consider following expriments:

```
template <typename T> T func(T t) { return t; }

template <typename T> void foo1(std::function<T(T)> f) {
  std::cout << f(0.1f) << std::endl;
}

template <typename T> void foo2(T (*f)(T)) {
  std::cout << f(0.1f) << std::endl;
}

void main(){
  foo1<float>([](float x)->float{ return x; }); // works
  foo1<float>([](int x)->int{ return x; });     // works, compiler failed to check type
  foo1<float>(func);                            // error, you have to specify func<float>
  foo1<float>(func<int>);                       // works, compiler failed to check type

  foo2<float>([](float x)->float{ return x; }); // works
  foo2<float>([](int x)->int{ return x; });     // error: good check
  foo2<float>(func);                            // works, compiler choose the correct template parameter for you
  foo2<float>(func<int>);                       // error: good check
}
```

The only limitation is T(*f)(T) cannot accept a lambda with capture, while std::function<T(T)> accepts any callable.
That's why I introduced another overload take any callable, but the same problem comes along, by allowing any callable, we lost the compile time signature checking, I don't know how to solve it yet.

Reply:
OK. so leave it as it is.